servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java

```diff
@@ -1534,11 +1543,39 @@ void deleteDomain(ResourceContext ctx, String auditRef, String domainName, Strin
             groupMemberConsistencyCheck(domainName, group.getName(), true, caller);
         }
 
-        // consistency checks are ok so we can go ahead and delete the domain
+        // consistency checks are ok. Now make sure no service is dependent on the domain
+
+        verifyNoServiceDependenciesOnDomain(domainName, caller);
+
+        // no service is dependent on the domain, we can go ahead and delete the domain
 
         dbService.executeDeleteDomain(ctx, domainName, auditRef, caller);
     }
 
+    private void verifyNoServiceDependenciesOnDomain(String domainName, String caller) {
+        ServiceIdentityList serviceIdentityList = dbService.listServiceDependencies(domainName);
+        if (serviceIdentityList.getNames() != null && !serviceIdentityList.getNames().isEmpty()) {
+            StringBuilder msgBuilder = new StringBuilder("Remove domain '");
+            msgBuilder.append(domainName);
+            msgBuilder.append("' dependency from the following service(s):");
+            msgBuilder.append(String.join(", ", serviceIdentityList.getNames()));
+            throw ZMSUtils.requestError(msgBuilder.toString(), caller);
+        }
+    }
+
+    private void verifyNoDomainDependenciesOnService(String serviceName, String caller) {
+        if (serviceProviderManager.isServiceProvider(serviceName)) {
```

Comment:
Just now realized I have a bug here.
I'll remove the line "if (serviceProviderManager.isServiceProvider(serviceName)) {" in another commit containing fixes to the other issues that will be raised in code review.
Explanation:
verifyNoDomainDependenciesOnService runs before we delete a service.
I'm verifying if the service is a service provider before checking for dependencies to prevent additional db calls.
This means that if we take a service provider and remove it from the service providers list (our sys.auth:role.service_providers role), we'll be able to delete it and this check will be ignored, possibly leaving "dangling" dependencies (dependencies on a service that no longer exist).